### PR TITLE
Provider type ID should be stored in creds.json (with backwards compatibility)

### PR DESCRIPTION
**Pull request by tlimoncelli**

### CODE REVIEW

**Comment by philpennock** [comment]
Not too fond of the magic `"-"` which is easy to overlook and will need lookups by users not familiar with the history, each time.  `"autoProvider"` might be better and will convey semantic meaning when seen.

Is the structure such that this can't just be dropped as mandatory and reflection used to decide if the next parameter is a provider type or an option?

**Comment by tlimoncelli** [comment]
Interesting point!

Looking over the work I've done so far, the reflection is working in all cases I've tried.  I guess I can just get rid of the whole "-" concept.  What I haven't tried so far is updating the command-line tools.  They have positional arguments but I might find some way to handle the old and new format.

Thanks for the feedback!  I'll look into this later this week.

**Comment by tlimoncelli** [comment]
> Not too fond of the magic `"-"` which is easy to overlook and will need lookups by users not familiar with the history, each time. `"autoProvider"` might be better and will convey semantic meaning when seen.

Updated as suggested.  The only place `-` is needed now is on the command line tools that use it as a positional argument.  Next step is to eliminate that last case when the positional argument is the last one on the command line. If that's all cases, then we can eliminate all the `-` shenanigans.

File: commands/previewPush.go
Comment (philpennock) [inline commands/previewPush.go:246]: specifies

File: pkg/js/helpers.js
Comment (philpennock) [inline pkg/js/helpers.js:48]: Is `break` after `return` a stylistic choice?  (My JS is poor so I'm a little confused)

**Review by philpennock — APPROVED** [review]

File: pkg/js/helpers.js
Comment (tlimoncelli) [reply pkg/js/helpers.js]: Nope.  Just a typo.

**Comment by tlimoncelli** [comment]
Good news, @philpennock !  While working through all the remaining code I was able to find ways to make everything forwards and backwards compatible.  From v3.16 on the new format will work.  In v4.0 we can remove the backwards compatibility.  All files and command line syntax works both ways, as long as people update their creds.json file to include a TYPE field.

**Comment by philpennock** [comment]
(double-check: you will merge when you're happy, since I've already approved and you're not using a workflow where the approver merges, right?)

**Comment by tlimoncelli** [comment]
Sadly I found a bug.  I'll work on this more later in the week. I have a big deadline Wed.

**Comment by tlimoncelli** [comment]
Bug fixed.

@philpennock and others: I'd appreciate some testing before I do this merge.

* Try the new binary with zero changes to your config. (dnscontrol preview is sufficient).
* Update your configuration based on the warnings and informational messages that are output.  Try "dnscontrol preview" and push.

Let me know if there are bugs, if anything is confusing, etc.

**Comment by tlimoncelli** [comment]
Closing.  All changes for this were moved to other PRs.

**Pull request closed without merging**
